cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll

```diff
@@ -1192,177 +1216,206 @@ private predicate nodeCand2(NodeExt node, boolean toReturn, boolean stored, Conf
  * Holds if `f` is the target of a read in the flow covered by `nodeCand2`.
  */
 pragma[noinline]
-private predicate readCand2(Content f, Configuration config) {
+private predicate readCand2(Content f, boolean read, Configuration config) {
   exists(NodeExt mid, NodeExt node |
     useFieldFlow(config) and
     nodeCandFwd2(node, _, true, unbind(config)) and
-    readExt(node, f, mid) and
-    storeCandFwd2(f, unbind(config)) and
-    nodeCand2(mid, _, _, config)
-  )
-}
-
-pragma[nomagic]
-private predicate nodeCand2Store(Content f, NodeExt node, boolean toReturn, Configuration config) {
-  exists(NodeExt mid |
-    storeExt(node, f, mid) and
-    nodeCand2(mid, toReturn, true, config)
+    readExt(node, f, mid, config) and
+    storeCandFwd2(f, unbindBool(read), unbind(config)) and
+    nodeCand2(mid, _, read, config)
   )
 }
 
 pragma[nomagic]
-private predicate nodeCand2TaintStore(
-  Content f, NodeExt node, boolean toReturn, Configuration config
+private predicate nodeCand2Store(
+  Content f, NodeExt node, boolean toReturn, boolean stored, Configuration config
 ) {
   exists(NodeExt mid |
-    argumentFlowsThroughFwd2(node, mid, TSummaryTaintStore(f), config) and
-    nodeCand2(mid, toReturn, true, config)
+    storeExt(node, f, mid, config) and
+    nodeCand2(mid, toReturn, true, config) and
+    nodeCandFwd2(node, _, stored, unbind(config))
   )
 }
 
 pragma[nomagic]
-private predicate storeCand(Content f, Configuration conf) {
+private predicate storeCand2(Content f, boolean stored, Configuration conf) {
   exists(NodeExt node |
-    nodeCand2Store(f, node, _, conf) or
-    nodeCand2TaintStore(f, node, _, conf)
-  |
-    nodeCand2(node, _, _, conf)
+    nodeCand2Store(f, node, _, stored, conf) and
+    nodeCand2(node, _, stored, conf)
   )
 }
 
 /**
  * Holds if `f` is the target of both a store and a read in the path graph
- * covered by `nodeCand2`.
+ * covered by `nodeCand2`. `apNonEmpty` indiciates whether some access path
+ * before the store (and after the read) is non-empty.
  */
 pragma[noinline]
-private predicate readStoreCand(Content f, Configuration conf) {
-  storeCand(f, conf) and
-  readCand2(f, conf)
+private predicate readStoreCand(Content f, boolean apNonEmpty, Configuration conf) {
```

Comment:
The column `apNonEmpty` isn't used, so should be moved into an `exists`.

Reply:
👍